packages/apiserver/src/resources/materials/material.service.ts

```diff
@@ -158,4 +126,362 @@ export class MaterialService {
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addExcludeLocationsCondition(
+        queryBuilder,
+        searchInput.excludeLocations,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addSearchCondition(queryBuilder, searchInput.search, isFirstWhere)
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addStatusesCondition(
+        queryBuilder,
+        searchInput.statuses,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addExcludeStatusesCondition(
+        queryBuilder,
+        searchInput.excludeStatuses,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addOverdueCondition(queryBuilder, searchInput.overdue, isFirstWhere)
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addAcceptedCondition(
+        queryBuilder,
+        searchInput.accepted,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addAuthorsCondition(queryBuilder, searchInput.authors, isFirstWhere)
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addExcludeAuthorsCondition(
+        queryBuilder,
+        searchInput.excludeAuthors,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (this.addTypesCondition(queryBuilder, searchInput.types, isFirstWhere)) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addExcludeTypesCondition(
+        queryBuilder,
+        searchInput.excludeTypes,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addCategoriesCondition(
+        queryBuilder,
+        searchInput.categories,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    if (
+      this.addExcludeCategoriesCondition(
+        queryBuilder,
+        searchInput.excludeCategories,
+        isFirstWhere
+      )
+    ) {
+      isFirstWhere = false;
+    }
+    return queryBuilder;
+  }
+
+  private addLocationsCondition(
+    queryBuilder: SelectQueryBuilder<Material>,
+    locations: number[],
+    isFirstWhere: boolean
+  ): boolean {
+    if (locations?.length) {
+      queryBuilder[isFirstWhere ? 'where' : 'andWhere'](
+        'material.location_id IN (:...locations)',
+        { locations }
+      );
+      return true;
+    }
+    return false;
+  }
+
+  private addExcludeLocationsCondition(
+    queryBuilder: SelectQueryBuilder<Material>,
+    excludeLocations: number[],
+    isFirstWhere: boolean
+  ): boolean {
+    if (excludeLocations?.length) {
+      queryBuilder[isFirstWhere ? 'where' : 'andWhere'](
+        'material.location_id NOT IN (:...excludeLocations)',
+        {
+          excludeLocations,
+        }
+      );
+      return true;
+    }
+    return false;
+  }
+
+  private addSearchCondition(
+    queryBuilder: SelectQueryBuilder<Material>,
+    search: string,
```

Comment:
can this variable be `undefined` if you check it in `if` statement? Better add `undefined` type here and below as it could be misleading

Reply:
Added undefined